packages/lms-client/src/llm/LLMDynamicHandle.ts

```diff
@@ -490,6 +736,422 @@ export class LLMDynamicHandle extends DynamicHandle<
+            case "success": {
+              const predictionResult = new PredictionResultWithRoundIndex(
+                contentArray.join(""),
+                message.stats,
+                message.modelInfo,
+                message.loadModelConfig,
+                message.predictionConfig,
+                predictionsPerformed,
+              );
+              safeCallCallback(
+                this.logger,
+                "onPredictionCompleted",
+                extraOpts.onPredictionCompleted,
+                [predictionResult],
+              );
+              predictionResolve();
+              break;
+            }
+          }
+        },
+        { stack },
+      );
+      const abortListener = () => {
+        if (finished) {
+          return;
+        }
+        finished = true;
+        channel.send({ type: "cancel" });
+      };
+      abortController.signal.addEventListener("abort", abortListener);
+      channel.onError.subscribeOnce(error => {
+        finished = true;
+        predictionReject(error);
+      });
+
+      predictionPromise
+        // When the prediction is completed, wait for all tool calls to be completed.
+        .then(() => Promise.all(toolCallPromises))
+        .then(() => finalResolve(), finalReject);
+
+      await finalPromise;
+
+      // Append and emit the assistant message.
+      const assistantMessage = ChatMessage.from({
+        role: "assistant",
+        content: [
+          {
+            type: "text",
+            text: contentArray.join(""),
+          },
+          ...toolCallRequests.map<ChatMessagePartToolCallRequestData>(toolCallRequest => ({
+            type: "toolCallRequest",
+            toolCallRequest,
+          })),
+        ],
+      });
+      mutableChat.append(assistantMessage.asMutableCopy());
+      safeCallCallback(this.logger, "onMessage", extraOpts.onMessage, [assistantMessage]);
+
+      shouldContinue = false;
+      if (toolCallResults.length > 0) {
+        // Sort the tool call results back into the order they were requested.
+        toolCallResults.sort((a, b) => a.index - b.index);
+
+        // Emit the tool call results.
+        const toolMessage = ChatMessage.from({
+          role: "tool",
+          content: toolCallResults.map(r => r.data),
+        });
+        mutableChat.append(toolMessage.asMutableCopy());
+        safeCallCallback(this.logger, "onMessage", extraOpts.onMessage, [toolMessage]);
+        shouldContinue = true;
+      }
+
+      safeCallCallback(this.logger, "onRoundEnd", extraOpts.onRoundEnd, [predictionsPerformed]);
+
+      predictionsPerformed++;
+      // Don't continue if we've reached the max predictions.
+      if (
+        extraOpts.maxPredictionRounds !== undefined &&
+        predictionsPerformed >= extraOpts.maxPredictionRounds
+      ) {
+        shouldContinue = false;
+      }
+    } while (shouldContinue);
+    return new OperationResult(predictionsPerformed, (performance.now() - startTime) / 1_000);
```

Comment:
@yagil This is the piece I'm looking to do differently in the Python SDK by adding a `predictions` sequence with the actual `PredictionResult[str]` from each round, rather than just counting the number of rounds and relying solely on the callbacks to publish the predictions themselves.

Python is *nowhere* near as callback heavy as Javascript, so it's rare to have an API where you *must* supply a callback to get useful info out of it.